File: packages/engine/src/engine/session_analysis.py
"""
Session analysis: Asia/London/NY session highs and lows, fair value gaps,
liquidity sweeps, and trade opportunity suggestions.
Uses 1h OHLC when available; falls back to daily.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

import pandas as pd
import yfinance as yf

# Session windows (UTC hours): Asia 00–08, London 08–16, NY 14–22 (open ~14:30)
ASIA_START, ASIA_END = 0, 8
LONDON_START, LONDON_END = 8, 16
NY_START, NY_END = 14, 22


@dataclass
class SessionLevels:
    name: str
    high: float
    low: float
    open: float
    close: float
    start_utc: str
    end_utc: str


@dataclass
class FairValueGap:
    kind: str  # "bullish" | "bearish"
    top: float
    bottom: float
    bar_index: int
    label: str = ""


@dataclass
class LiquiditySweep:
    kind: str  # "high_sweep" | "low_sweep"
    level: float
    bar_index: int
    label: str = ""


@dataclass
class SessionAnalysis:
    symbol: str
    session_levels: list[SessionLevels] = field(default_factory=list)
    fvgs: list[FairValueGap] = field(default_factory=list)
    sweeps: list[LiquiditySweep] = field(default_factory=list)
    predicting_ny: str = ""
    opportunities: list[str] = field(default_factory=list)
    daily_high: float = 0.0
    daily_low: float = 0.0
    daily_open: float = 0.0
    daily_close: float = 0.0
    daily_change_pct: float = 0.0
    pattern: str = ""
    pattern_color: str = "dim"
    date: str = ""
    interval: str = "1h"  # "1h" | "1d" (what we used)


def _get_1h(symbol: str, period: str = "5d") -> pd.DataFrame | None:
    try:
        t = yf.Ticker(symbol)
        df = t.history(period=period, interval="1h")
        if df is None or df.empty or len(df) < 3:
            return None
        # Ensure timezone-aware UTC
        if df.index.tz is None:
            df.index = df.index.tz_localize("America/New_York", ambiguous="infer").tz_convert("UTC")
        else:
            df.index = df.index.tz_convert("UTC")
        return df
    except Exception:
        return None


def _get_daily(symbol: str, period: str = "5d") -> pd.DataFrame | None:
    try:
        t = yf.Ticker(symbol)
        df = t.history(period=period, interval="1d")
        if df is None or df.empty or len(df) < 2:
            return None
        return df
    except Exception:
        return None


def _session_name(hour_utc: int) -> str:
    if ASIA_START <= hour_utc < ASIA_END:
        return "Asia"
    if LONDON_START <= hour_utc < LONDON_END:
        return "London"
    if NY_START <= hour_utc < NY_END:
        return "NY"
    return "Overnight"


def _aggregate_session_levels(df: pd.DataFrame) -> list[SessionLevels]:
    """Bucket 1h bars by session and compute high/low/open/close per session."""
    out: list[SessionLevels] = []
    df = df.copy()
    df["hour"] = df.index.hour
    df["session"] = df["hour"].apply(_session_name)
    df["date"] = df.index.date

    for (date, session_name), grp in df.groupby(["date", "session"]):
        if session_name == "Overnight":
            continue
        high = float(grp["High"].max())
        low = float(grp["Low"].min())
        open_ = float(grp["Open"].iloc[0])
        close = float(grp["Close"].iloc[-1])
        start_utc = grp.index.min().strftime("%Y-%m-%d %H:%M") if hasattr(grp.index.min(), "strftime") else str(grp.index.min())
        end_utc = grp.index.max().strftime("%Y-%m-%d %H:%M") if hasattr(grp.index.max(), "strftime") else str(grp.index.max())
        out.append(SessionLevels(
            name=session_name,
            high=high,
            low=low,
            open=open_,
            close=close,
            start_utc=start_utc,
            end_utc=end_utc,
        ))
    return out[-12:]  # last ~12 sessions (e.g. 4 days of 3 sessions)


def _detect_fvg(df: pd.DataFrame, lookback: int = 48) -> list[FairValueGap]:
    """Fair value gaps: 3-candle pattern. Bullish = low[i+2] > high[i]; Bearish = high[i+2] < low[i]."""
    fvgs: list[FairValueGap] = []
    arr = df.tail(lookback)
    if len(arr) < 3:
        return fvgs
    highs = arr["High"].values
    lows = arr["Low"].values
    for i in range(len(arr) - 2):
        if lows[i + 2] > highs[i]:
            fvgs.append(FairValueGap(
                kind="bullish",
                top=float(lows[i + 2]),
                bottom=float(highs[i]),
                bar_index=i + 2,
                label=f"Bullish FVG {highs[i]:.2f}–{lows[i+2]:.2f}",
            ))
        if highs[i + 2] < lows[i]:
            fvgs.append(FairValueGap(
                kind="bearish",
                top=float(lows[i]),
                bottom=float(highs[i + 2]),
                bar_index=i + 2,
                label=f"Bearish FVG {highs[i+2]:.2f}–{lows[i]:.2f}",
            ))
    return fvgs[-6:]  # keep last 6


def _detect_sweeps(df: pd.DataFrame, lookback: int = 24, n_prior: int = 5) -> list[LiquiditySweep]:
    """Liquidity sweep: price takes out prior N-bar high/low then closes back inside."""
    sweeps: list[LiquiditySweep] = []
    arr = df.tail(lookback)
    if len(arr) < n_prior + 2:
        return sweeps
    highs = arr["High"].values
    lows = arr["Low"].values
    opens = arr["Open"].values
    closes = arr["Close"].values

    for i in range(n_prior, len(arr) - 1):
        prior_high = float(max(highs[i - n_prior : i]))
        prior_low = float(min(lows[i - n_prior : i]))
        h, l, o, c = highs[i], lows[i], opens[i], closes[i]
        # High sweep: breaks above prior high then closes below prior high (or in lower half of bar)
        if h > prior_high and c < prior_high:
            sweeps.append(LiquiditySweep(
                kind="high_sweep",
                level=prior_high,
                bar_index=i,
                label=f"Highs swept at {prior_high:.2f}",
            ))
        if l < prior_low and c > prior_low:
            sweeps.append(LiquiditySweep(
                kind="low_sweep",
                level=prior_low,
                bar_index=i,
                label=f"Lows swept at {prior_low:.2f}",
            ))
    return sweeps[-4:]  # last 4


def _predicting_ny(sessions: list[SessionLevels], now_utc: datetime) -> str:
    """Plain-text summary for 'predicting in NY' (CLI adds color when rendering)."""
    if not sessions:
        return ""
    hour = now_utc.hour
    recent = [s for s in sessions[-6:] if s.name in ("Asia", "London", "NY")]
    if not recent:
        return ""
    asia = next((s for s in reversed(recent) if s.name == "Asia"), None)
    london = next((s for s in reversed(recent) if s.name == "London"), None)
    ny = next((s for s in reversed(recent) if s.name == "NY"), None)

    parts = []
    if hour < NY_START and asia:
        parts.append(f"Asia high {asia.high:.2f} · Asia low {asia.low:.2f}")
    if hour < NY_START and london:
        parts.append(f"London high {london.high:.2f} · London low {london.low:.2f}")
    if hour < NY_START:
        parts.append("→ Watch for reaction at NY open (14:30 UTC)")
    elif ny:
        parts.append(f"NY session high {ny.high:.2f} · NY low {ny.low:.2f}")
    return " · ".join(parts) if parts else ""


def _build_opportunities(
    sessions: list[SessionLevels],
    fvgs: list[FairValueGap],
    sweeps: list[LiquiditySweep],
    daily_high: float,
    daily_low: float,
    daily_close: float,
) -> list[str]:
    opps: list[str] = []
    recent = sessions[-6:] if sessions else []

    for fvg in fvgs[-3:]:
        if fvg.kind == "bullish" and daily_close < fvg.top:
            opps.append(f"Long opportunity: bullish FVG {fvg.bottom:.2f}–{fvg.top:.2f} — look for long on retest")
        elif fvg.kind == "bearish" and daily_close > fvg.bottom:
            opps.append(f"Short opportunity: bearish FVG {fvg.bottom:.2f}–{fvg.top:.2f} — look for short on retest")

    for sw in sweeps[-2:]:
        if sw.kind == "low_sweep":
            opps.append(f"Liquidity sweep (lows at {sw.level:.2f}) — potential long if structure holds")
        else:
            opps.append(f"Liquidity sweep (highs at {sw.level:.2f}) — potential short if structure holds")

    asia = next((s for s in reversed(recent) if s.name == "Asia"), None)
    london = next((s for s in reversed(recent) if s.name == "London"), None)
    ny = next((s for s in reversed(recent) if s.name == "NY"), None)
    if asia and daily_close <= asia.high and daily_close >= asia.low:
        opps.append(f"Price inside Asia range {asia.low:.2f}–{asia.high:.2f} — use as support/resistance")
    if london and daily_close <= london.high and daily_close >= london.low:
        opps.append(f"Price inside London range {london.low:.2f}–{london.high:.2f} — key level for NY")
    if ny and daily_close <= ny.high and daily_close >= ny.low:
        opps.append(f"NY range {ny.low:.2f}–{ny.high:.2f} — hold for trend or break for next target")

    return opps[:6]  # cap at 6


def _detect_pattern(df: pd.DataFrame) -> tuple[str, str]:
    if df is None or len(df) < 3:
        return "n/a", "dim"
    highs = df["High"].values
    lows = df["Low"].values
    h1, h0 = highs[-1], highs[-2]
    l1, l0 = lows[-1], lows[-2]
    if h1 > h0 and l1 > l0:
        return "HH/HL (bullish)", "bold green"
    if h1 < h0 and l1 < l0:
        return "LH/LL (bearish)", "bold red"
    if h1 > h0 and l1 < l0:
        return "HH/LL (expansion)", "bold yellow1"
    return "LH/HL (contraction)", "bold blue1"


def run_session_analysis(symbol: str) -> SessionAnalysis | None:
    """Run full session analysis: 1h session levels, FVGs, sweeps, predicting NY, opportunities."""
    df_1h = _get_1h(symbol, period="5d")
    df_daily = _get_daily(symbol, period="5d")
    if df_daily is None or df_daily.empty:
        return None

    last = df_daily.iloc[-1]
    daily_high = float(last["High"])
    daily_low = float(last["Low"])
    daily_open = float(last["Open"])
    daily_close = float(last["Close"])
    daily_change_pct = ((daily_close - daily_open) / daily_open * 100) if daily_open else 0.0
    pattern, pattern_color = _detect_pattern(df_daily)
    date_str = str(df_daily.index[-1].date()) if hasattr(df_daily.index[-1], "date") else str(df_daily.index[-1])

    session_levels: list[SessionLevels] = []
    fvgs: list[FairValueGap] = []
    sweeps: list[LiquiditySweep] = []
    interval = "1d"
    predicting_ny = ""
    opportunities: list[str] = []

    if df_1h is not None and len(df_1h) >= 3:
        interval = "1h"
        session_levels = _aggregate_session_levels(df_1h)
        fvgs = _detect_fvg(df_1h)
        sweeps = _detect_sweeps(df_1h)
        now = datetime.now(timezone.utc)
        predicting_ny = _predicting_ny(session_levels, now)
        opportunities = _build_opportunities(
            session_levels, fvgs, sweeps,
            daily_high, daily_low, daily_close,
        )

    return SessionAnalysis(
        symbol=symbol,
        session_levels=session_levels,
        fvgs=fvgs,
        sweeps=sweeps,
        predicting_ny=predicting_ny,
        opportunities=opportunities,
        daily_high=daily_high,
        daily_low=daily_low,
        daily_open=daily_open,
        daily_close=daily_close,
        daily_change_pct=daily_change_pct,
        pattern=pattern,
        pattern_color=pattern_color,
        date=date_str,
        interval=interval,
    )


def analysis_to_dict(a: SessionAnalysis) -> dict[str, Any]:
    """Serialize for API/JSON."""
    return {
        "symbol": a.symbol,
        "date": a.date,
        "interval": a.interval,
        "daily_high": a.daily_high,
        "daily_low": a.daily_low,
        "daily_open": a.daily_open,
        "daily_close": a.daily_close,
        "daily_change_pct": a.daily_change_pct,
        "pattern": a.pattern,
        "pattern_color": a.pattern_color,
        "session_levels": [
            {
                "name": s.name,
                "high": s.high,
                "low": s.low,
                "open": s.open,
                "close": s.close,
                "start_utc": s.start_utc,
                "end_utc": s.end_utc,
            }
            for s in a.session_levels
        ],
        "fvgs": [
            {"kind": f.kind, "top": f.top, "bottom": f.bottom, "label": f.label}
            for f in a.fvgs
        ],
        "sweeps": [
            {"kind": s.kind, "level": s.level, "label": s.label}
            for s in a.sweeps
        ],
        "predicting_ny": a.predicting_ny,
        "opportunities": a.opportunities,
    }

File: apps/web/lib/sessionAnalysis.ts
/**
 * Session analysis: Asia/London/NY session levels, fair value gaps,
 * liquidity sweeps, and trade opportunities. Uses 1h candles (UTC hours).
 */
import type { Candle } from "./market";

const ASIA_START = 0;
const ASIA_END = 8;
const LONDON_START = 8;
const LONDON_END = 16;
const NY_START = 14;
const NY_END = 22;

export type SessionLevel = {
  name: string;
  high: number;
  low: number;
  open: number;
  close: number;
  start_utc: string;
  end_utc: string;
};

export type FairValueGap = {
  kind: "bullish" | "bearish";
  top: number;
  bottom: number;
  label: string;
};

export type LiquiditySweep = {
  kind: "high_sweep" | "low_sweep";
  level: number;
  label: string;
};

/** What price does at the same time every day (by UTC hour). */
export type TimeOfDayBucket = {
  hourUtc: number;
  label: string;
  days: number;
  avgReturnPct: number;
  winRate: number;
  avgRangePct: number;
};

/** Exact trade to take: time, entry, SL, TP (auto-calculated for best R:R from structure). */
export type TradeSuggestion = {
  timeLabel: string;
  side: "long" | "short";
  entry: number;
  sl: number;
  tp: number;
  rr: number;
  reason: string;
};

export type SessionAnalysisResult = {
  session_levels: SessionLevel[];
  fvgs: FairValueGap[];
  sweeps: LiquiditySweep[];
  predicting_ny: string;
  opportunities: string[];
  time_of_day?: TimeOfDayBucket[];
  trades?: TradeSuggestion[];
};

function sessionName(hourUtc: number): string {
  if (hourUtc >= ASIA_START && hourUtc < ASIA_END) return "Asia";
  if (hourUtc >= LONDON_START && hourUtc < LONDON_END) return "London";
  if (hourUtc >= NY_START && hourUtc < NY_END) return "NY";
  return "Overnight";
}

/** Assume candle times are in seconds (epoch). Convert to UTC for bucketing. */
function aggregateSessionLevels(candles: Candle[]): SessionLevel[] {
  const buckets: Map<string, { high: number; low: number; opens: number[]; closes: number[]; times: number[] }> = new Map();
  for (const c of candles) {
    const d = new Date(c.time * 1000);
    const date = d.toISOString().slice(0, 10);
    const hour = d.getUTCHours();
    const name = sessionName(hour);
    if (name === "Overnight") continue;
    const key = `${date}:${name}`;
    let b = buckets.get(key);
    if (!b) {
      b = { high: c.high, low: c.low, opens: [], closes: [], times: [] };
      buckets.set(key, b);
    }
    b.high = Math.max(b.high, c.high);
    b.low = Math.min(b.low, c.low);
    b.opens.push(c.open);
    b.closes.push(c.close);
    b.times.push(c.time);
  }
  const out: SessionLevel[] = [];
  const keys = Array.from(buckets.keys()).sort();
  for (const key of keys) {
    const b = buckets.get(key)!;
    const [date, name] = key.split(":");
    const sortedTimes = [...b.times].sort((a, b) => a - b);
    const startUtc = new Date(sortedTimes[0] * 1000).toISOString().slice(0, 16).replace("T", " ");
    const endUtc = new Date(sortedTimes[sortedTimes.length - 1] * 1000).toISOString().slice(0, 16).replace("T", " ");
    out.push({
      name,
      high: b.high,
      low: b.low,
      open: b.opens[0],
      close: b.closes[b.closes.length - 1],
      start_utc: startUtc,
      end_utc: endUtc,
    });
  }
  return out.slice(-12);
}

function detectFvg(candles: Candle[], lookback = 48): FairValueGap[] {
  const fvgs: FairValueGap[] = [];
  const arr = candles.slice(-lookback);
  if (arr.length < 3) return fvgs;
  for (let i = 0; i < arr.length - 2; i++) {
    const high0 = arr[i].high;
    const low0 = arr[i].low;
    const high2 = arr[i + 2].high;
    const low2 = arr[i + 2].low;
    if (low2 > high0) {
      fvgs.push({
        kind: "bullish",
        top: low2,
        bottom: high0,
        label: `Bullish FVG ${high0.toFixed(2)}–${low2.toFixed(2)}`,
      });
    }
    if (high2 < low0) {
      fvgs.push({
        kind: "bearish",
        top: low0,
        bottom: high2,
        label: `Bearish FVG ${high2.toFixed(2)}–${low0.toFixed(2)}`,
      });
    }
  }
  return fvgs.slice(-6);
}

function detectSweeps(candles: Candle[], lookback = 24, nPrior = 5): LiquiditySweep[] {
  const sweeps: LiquiditySweep[] = [];
  const arr = candles.slice(-lookback);
  if (arr.length < nPrior + 2) return sweeps;
  for (let i = nPrior; i < arr.length - 1; i++) {
    let priorHigh = arr[i - 1].high;
    let priorLow = arr[i - 1].low;
    for (let j = i - nPrior; j < i; j++) {
      priorHigh = Math.max(priorHigh, arr[j].high);
      priorLow = Math.min(priorLow, arr[j].low);
    }
    const h = arr[i].high;
    const l = arr[i].low;
    const c = arr[i].close;
    if (h > priorHigh && c < priorHigh) {
      sweeps.push({ kind: "high_sweep", level: priorHigh, label: `Highs swept at ${priorHigh.toFixed(2)}` });
    }
    if (l < priorLow && c > priorLow) {
      sweeps.push({ kind: "low_sweep", level: priorLow, label: `Lows swept at ${priorLow.toFixed(2)}` });
    }
  }
  return sweeps.slice(-4);
}

function predictingNy(sessions: SessionLevel[], nowUtc: Date): string {
  const hour = nowUtc.getUTCHours();
  const recent = sessions.filter((s) => s.name === "Asia" || s.name === "London" || s.name === "NY").slice(-6);
  if (!recent.length) return "";
  const asia = [...recent].reverse().find((s) => s.name === "Asia");
  const london = [...recent].reverse().find((s) => s.name === "London");
  const ny = [...recent].reverse().find((s) => s.name === "NY");
  const parts: string[] = [];
  if (hour < NY_START && asia) {
    parts.push(`Asia high ${asia.high.toFixed(2)} · Asia low ${asia.low.toFixed(2)}`);
  }
  if (hour < NY_START && london) {
    parts.push(`London high ${london.high.toFixed(2)} · London low ${london.low.toFixed(2)}`);
  }
  if (hour < NY_START) {
    parts.push("→ Watch for reaction at NY open (14:30 UTC)");
  } else if (ny) {
    parts.push(`NY session high ${ny.high.toFixed(2)} · NY low ${ny.low.toFixed(2)}`);
  }
  return parts.join(" · ");
}

function buildOpportunities(
  sessions: SessionLevel[],
  fvgs: FairValueGap[],
  sweeps: LiquiditySweep[],
  dailyHigh: number,
  dailyLow: number,
  dailyClose: number
): string[] {
  const opps: string[] = [];
  const recent = sessions.slice(-6);
  for (const fvg of fvgs.slice(-3)) {
    if (fvg.kind === "bullish" && dailyClose < fvg.top) {
      opps.push(`Long: bullish FVG ${fvg.bottom.toFixed(2)}–${fvg.top.toFixed(2)} — look for long on retest`);
    } else if (fvg.kind === "bearish" && dailyClose > fvg.bottom) {
      opps.push(`Short: bearish FVG ${fvg.bottom.toFixed(2)}–${fvg.top.toFixed(2)} — look for short on retest`);
    }
  }
  for (const sw of sweeps.slice(-2)) {
    if (sw.kind === "low_sweep") {
      opps.push(`Liquidity sweep (lows at ${sw.level.toFixed(2)}) — potential long if structure holds`);
    } else {
      opps.push(`Liquidity sweep (highs at ${sw.level.toFixed(2)}) — potential short if structure holds`);
    }
  }
  const asia = [...recent].reverse().find((s) => s.name === "Asia");
  const london = [...recent].reverse().find((s) => s.name === "London");
  const ny = [...recent].reverse().find((s) => s.name === "NY");
  if (asia && dailyClose <= asia.high && dailyClose >= asia.low) {
    opps.push(`Price inside Asia range ${asia.low.toFixed(2)}–${asia.high.toFixed(2)} — use as S/R`);
  }
  if (london && dailyClose <= london.high && dailyClose >= london.low) {
    opps.push(`Price inside London range ${london.low.toFixed(2)}–${london.high.toFixed(2)} — key level for NY`);
  }
  if (ny && dailyClose <= ny.high && dailyClose >= ny.low) {
    opps.push(`NY range ${ny.low.toFixed(2)}–${ny.high.toFixed(2)} — hold for trend or break for target`);
  }
  return opps.slice(0, 6);
}

/** Group 1h candles by (date, UTC hour). For each hour: days count, avg return %, win rate, avg range %. */
function computeTimeOfDayStats(candles1h: Candle[]): TimeOfDayBucket[] {
  const byHour = new Map<
    number,
    { returns: number[]; ranges: number[] }
  >();
  let lastDate = "";
  const seenDateHour = new Set<string>();
  for (const c of candles1h) {
    const d = new Date(c.time * 1000);
    const date = d.toISOString().slice(0, 10);
    const hour = d.getUTCHours();
    const key = `${date}:${hour}`;
    if (seenDateHour.has(key)) continue;
    seenDateHour.add(key);
    if (c.open <= 0) continue;
    const retPct = ((c.close - c.open) / c.open) * 100;
    const rangePct = ((c.high - c.low) / c.open) * 100;
    let b = byHour.get(hour);
    if (!b) {
      b = { returns: [], ranges: [] };
      byHour.set(hour, b);
    }
    b.returns.push(retPct);
    b.ranges.push(rangePct);
  }
  const sessionHours = [0, 1, 7, 8, 9, 14, 15, 16, 20, 21];
  const labels: Record<number, string> = {
    0: "00:00 Asia",
    1: "01:00 Asia",
    7: "07:00 London pre",
    8: "08:00 London open",
    9: "09:00 London",
    14: "14:00 NY open",
    15: "15:00 NY",
    16: "16:00 NY",
    20: "20:00 NY close",
    21: "21:00 After hours",
  };
  const out: TimeOfDayBucket[] = [];
  for (const hour of sessionHours) {
    const b = byHour.get(hour);
    if (!b || b.returns.length < 2) continue;
    const n = b.returns.length;
    const avgReturnPct = b.returns.reduce((a, x) => a + x, 0) / n;
    const winRate = b.returns.filter((r) => r > 0).length / n;
    const avgRangePct = b.ranges.reduce((a, x) => a + x, 0) / n;
    out.push({
      hourUtc: hour,
      label: labels[hour] ?? `${hour}:00 UTC`,
      days: n,
      avgReturnPct,
      winRate,
      avgRangePct,
    });
  }
  return out.sort((a, b) => a.hourUtc - b.hourUtc);
}

const TARGET_RR = 1.5;
const SL_BUFFER_PCT = 0.08;

/** Build exact trades with auto SL/TP from structure (FVGs, session levels, sweeps). */
function buildTradeSuggestions(
  sessions: SessionLevel[],
  fvgs: FairValueGap[],
  sweeps: LiquiditySweep[],
  dailyHigh: number,
  dailyLow: number,
  dailyClose: number,
  timeOfDay: TimeOfDayBucket[]
): TradeSuggestion[] {
  const recent = sessions.slice(-6);
  const asia = [...recent].reverse().find((s) => s.name === "Asia");
  const london = [...recent].reverse().find((s) => s.name === "London");
  const ny = [...recent].reverse().find((s) => s.name === "NY");
  const suggestions: TradeSuggestion[] = [];

  const bestLongHour = [...timeOfDay].filter((t) => t.avgReturnPct > 0).sort((a, b) => b.winRate - a.winRate)[0];
  const bestShortHour = [...timeOfDay].filter((t) => t.avgReturnPct < 0).sort((a, b) => b.winRate - a.winRate)[0];
  const timeLabelLong = bestLongHour ? bestLongHour.label : "14:00 NY open";
  const timeLabelShort = bestShortHour ? bestShortHour.label : "14:00 NY open";

  for (const fvg of fvgs.slice(-3)) {
    if (fvg.kind === "bullish" && dailyClose <= fvg.top) {
      const entry = fvg.bottom;
      const risk = Math.max((fvg.top - fvg.bottom) * 0.5, entry * (SL_BUFFER_PCT / 100));
      const sl = entry - risk;
      const tp = entry + risk * TARGET_RR;
      const tpCapped = ny ? Math.min(tp, ny.high * 1.001) : tp;
      const rr = (tpCapped - entry) / (entry - sl);
      suggestions.push({
        timeLabel: timeLabelLong,
        side: "long",
        entry,
        sl,
        tp: tpCapped,
        rr,
        reason: `Bullish FVG retest ${fvg.bottom.toFixed(2)}–${fvg.top.toFixed(2)}; pattern at this time`,
      });
    } else if (fvg.kind === "bearish" && dailyClose >= fvg.bottom) {
      const entry = fvg.top;
      const risk = Math.max((fvg.top - fvg.bottom) * 0.5, entry * (SL_BUFFER_PCT / 100));
      const sl = entry + risk;
      const tp = entry - risk * TARGET_RR;
      const tpCapped = ny ? Math.max(tp, ny.low * 0.999) : tp;
      const rr = (entry - tpCapped) / (sl - entry);
      suggestions.push({
        timeLabel: timeLabelShort,
        side: "short",
        entry,
        sl,
        tp: tpCapped,
        rr,
        reason: `Bearish FVG retest ${fvg.bottom.toFixed(2)}–${fvg.top.toFixed(2)}; pattern at this time`,
      });
    }
  }

  for (const sw of sweeps.slice(-2)) {
    if (sw.kind === "low_sweep" && dailyClose > sw.level) {
      const entry = sw.level * 1.002;
      const risk = (london ? Math.min(entry - london.low, entry * 0.003) : entry * 0.003) || entry * 0.002;
      const sl = entry - risk;
      const tp = entry + risk * TARGET_RR;
      suggestions.push({
        timeLabel: timeLabelLong,
        side: "long",
        entry,
        sl,
        tp,
        rr: TARGET_RR,
        reason: `Long after lows swept at ${sw.level.toFixed(2)}; same-time behavior supports bounce`,
      });
    } else if (sw.kind === "high_sweep" && dailyClose < sw.level) {
      const entry = sw.level * 0.998;
      const risk = (london ? Math.min(london.high - entry, entry * 0.003) : entry * 0.003) || entry * 0.002;
      const sl = entry + risk;
      const tp = entry - risk * TARGET_RR;
      suggestions.push({
        timeLabel: timeLabelShort,
        side: "short",
        entry,
        sl,
        tp,
        rr: TARGET_RR,
        reason: `Short after highs swept at ${sw.level.toFixed(2)}; same-time behavior supports drop`,
      });
    }
  }

  if (asia && dailyClose >= asia.low && dailyClose <= asia.high && suggestions.length < 3) {
    const mid = (asia.low + asia.high) / 2;
    const entry = dailyClose;
    const risk = Math.max(entry - asia.low, entry * 0.002);
    const sl = entry - risk;
    const tp = entry + risk * TARGET_RR;
    suggestions.push({
      timeLabel: timeLabelLong,
      side: "long",
      entry,
      sl,
      tp: Math.min(tp, asia.high),
      rr: TARGET_RR,
      reason: `Long from Asia range ${asia.low.toFixed(2)}–${asia.high.toFixed(2)}; bounce at this time`,
    });
  }

  return suggestions.slice(0, 5);
}

export function runSessionAnalysis(
  candles1h: Candle[],
  dailyHigh: number,
  dailyLow: number,
  dailyClose: number
): SessionAnalysisResult {
  const session_levels = aggregateSessionLevels(candles1h);
  const fvgs = detectFvg(candles1h);
  const sweeps = detectSweeps(candles1h);
  const predicting_ny = predictingNy(session_levels, new Date());
  const opportunities = buildOpportunities(session_levels, fvgs, sweeps, dailyHigh, dailyLow, dailyClose);
  const time_of_day = computeTimeOfDayStats(candles1h);
  const trades = buildTradeSuggestions(
    session_levels,
    fvgs,
    sweeps,
    dailyHigh,
    dailyLow,
    dailyClose,
    time_of_day
  );
  return {
    session_levels,
    fvgs,
    sweeps,
    predicting_ny,
    opportunities,
    time_of_day,
    trades,
  };
}
